### ISSUE TITLE
Pin Code not working Packetver: 20130618

### ISSUE THREAD

**Issue by AtomixRO** [MrAntares/roBrowserLegacy#448]
**Describe the bug**
- Feature gets you straight to Verify Pin instead of Password for old accounts that works.

- Also Pin code in console is wrong compared to the pin code which has been typed but doesn't mention it's a wrong pincode.

- Change password gets stuck on Password & Confirm Password, can't get to "New Password" field.

- Korean button shouldn't be displayed on this packet version.

**Screenshots (if applicable)**
Typing pincode after login:
![image](https://github.com/user-attachments/assets/afc5d6fc-2e4f-40e8-b9b2-f0820b125b4b)

Change Password:
![image](https://github.com/user-attachments/assets/6a5beb84-456f-4653-817d-a73f11c20020)

Incorrect pincode pinted and I get an error for wrong pincode if I type the wrong one.
![image](https://github.com/user-attachments/assets/ee27def1-a3c8-4975-9540-b2a6689b5d5e)

**Packet/Client version**
20130618

**Browser/device info**
 - Any

**Comment by MrAntares** [comment]
@redpolline Do you also have the same issues on your packetver? I can't test it on my server since it is too old.

**Comment by redpolline** [comment]
Sorry for not getting to this until now. I've been away all week.

> * Feature gets you straight to Verify Pin instead of Password for old accounts that works.

> **Packet/Client version** 20130618
Your PACKETVER is old enough to be affected by a TODO.

Basically, I didn't know what one of the packets for pincodes did as it was deprecated before my PACKETVER. (rAthena won't send it in my case.)

Can you send the full log? What we really need is the received packets from the server and the order they were sent in. 

> * Also Pin code in console is wrong compared to the pin code which has been typed but doesn't mention it's a wrong pincode.

The pincode it prints in the console is the encrypted pin code.

If you know the userSeed that was sent, (check the logs), you can decrypt it back into the user's original pincode. (Although, you should be able to tell that it's your pin code. Especially if you have two digits in your pin that are the same. As they'll be the same digits in the encrypted version.)

> * Change password gets stuck on Password & Confirm Password, can't get to "New Password" field.

I may have screwed up the order for what password box is supposed to be used.... That should just be cosmetic though. It should still work if you type in the expected code.

> * Korean button shouldn't be displayed on this packet version.

I'm not sure what that button is supposed to be. Maybe it's the unused Exit button? (That should be hidden automatically though....) Can anyone translate it?

**Comment by redpolline** [comment]
> @redpolline Do you also have the same issues on your packetver? I can't test it on my server since it is too old.

Unfortunately, my server is too new.

The screenshots indicate that login packet state 7 was sent by the server. Which according to rAthena was removed in PACKETVER >= 20180124.

**Comment by redpolline** [comment]
> > * Korean button shouldn't be displayed on this packet version.
> I'm not sure what that button is supposed to be. Maybe it's the unused Exit button? (That should be hidden automatically though....) Can anyone translate it?

OK, that's the unused exit button. Your client files just don't have that button translated for some reason. (Out of date?)

I'll send a patch to disable it on all packetvers, as it's not wired up to anything to begin with.

**Comment by redpolline** [comment]
@AtomixRO Did the recent commit fix your issue?

**Comment by AtomixRO** [comment]
@redpolline Sorry for the long wait, been too busy.
- Login works.

- Change doesn't work:
When you click on Change, the box remains on Confirm Password: 
![image](https://github.com/user-attachments/assets/bcbe1107-e5a5-480a-8297-50b7c0ce10cc)
You click on Verify, it then goes to Password.
The Change Password procedure on client is:
After click on Change, you have to type your current PIN
![image](https://github.com/user-attachments/assets/9298dcf0-8b06-4f77-a22b-e03771ff971d)
And then you click on Verify, New Password is now accessible:
![image](https://github.com/user-attachments/assets/aa2a09f2-5029-4d3b-a1ac-d04348423c3c)
Click on Verify again and then Confirm Password is accessible and Confirm button appears:
![image](https://github.com/user-attachments/assets/916f1e9e-3f00-464d-8182-0cd9f43074b6)
When all information are filled in, if you click on Reset, The procedure will revert back to Password:
![image](https://github.com/user-attachments/assets/7d88a90f-60d0-4f7d-9ef1-992b94cbaf71)


- NEW PIN Code:
It's buggy but works for new accounts.
On client it's just like change minus to enter the current PIN code yet the box is white, not blue:
Enter New Password and then click on Verify.
![image](https://github.com/user-attachments/assets/cb992be4-a039-4745-82c5-68829c54969b)
Enter Confirm Password and then click on Confirm.
![image](https://github.com/user-attachments/assets/bb731321-3f75-4028-ad2d-353dc8e3fe5b)

Browser also get 3 notifications box overlapped: 
![image](https://github.com/user-attachments/assets/162cc280-de0e-40c5-8575-7fd236955378)

**Comment by MrAntares** [comment]
Same issue here. Can't make new pincode and can't login either. This original ui is pretty unintuitive with the lower case "change" to begin with, but the random fileds activating and the 3 popups make it a real a chaos :D I don't even know what I am supposed to do :D

**Comment by AtomixRO** [comment]
> Same issue here. Can't make new pincode and can't login either. This original ui is pretty unintuitive with the lower case "change" to begin with, but the random fileds activating and the 3 popups make it a real a chaos :D I don't even know what I am supposed to do :D

Click on OK 3 times in a row, type your new pin code and then click on Confirm, done. :D
That was my experience twice on browser.

**Comment by issid** [comment]
the problem is if you use the first digit 0 . it is removed in the code here `intCode = Number.parseInt(pincode);`  that's why we need to add a check
```                // add missing zero at the beginning
                if (strCode.length < pincode.length) {
                    var zeros = pincode.length - strCode.length;
                    for (var j = 0; j < zeros; j++) {
                        strCode = '0' + strCode;
                    }
                }
```
before the encryption code
```
                // Encrypt raw digits with pad.
                for (i = 0; i < strCode.length; i++) {
                    x = Number(strCode[i]);
                    out += (PincodeWindow._keypad[x]).toString();
                }
```
in the file `PincodeWindow.js`

**Comment by redpolline** [comment]
> @redpolline Sorry for the long wait, been too busy.
> 
> * Login works but on client the you should type your pin code in Password box
>   ![image](https://private-user-images.githubusercontent.com/214579/364837159-90006fce-5ea4-41b7-8661-58782f3758b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODM3MTU5LTkwMDA2ZmNlLTVlYTQtNDFiNy04NjYxLTU4NzgyZjM3NThiMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNmJkNGJlMTdjODU5Njk2NjUzNzUyMGNkMzYzZmM4MDE0ODQ1MzIzYjM1YTAyZTBkNGNiZTBiMzAyNGI4YzhlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.AlXF_VjIAkTtN2DVUAesA99Epl9Tim0dqdREdK0apbE)
>   Not in Confirm Password on browser:
>   ![image](https://private-user-images.githubusercontent.com/214579/364837717-321fc35b-c4e8-4761-a039-f4d985018f6c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODM3NzE3LTMyMWZjMzViLWM0ZTgtNDc2MS1hMDM5LWY0ZDk4NTAxOGY2Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZDFiNjJhMDdkODZhNjg0OTZlM2NkMzlmOTQwNDQwMWRmMWMzM2E2ZTRmNjI3MzYyNjZjNGVjMzYwYjc4YmUxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.cWp43_sEwM7oIxkhy3YMWiqTlFKFfw_yHYZ0sTRsGDs)

Yeah, that's just a cosmetic bug.... I didn't realize it when I wrote the code for setting the values from the buttons.

Basically, "Confirm Password" and "Password" are switched in the code. The data put into them (incorrectly) still gets used properly, but it does look wrong from the user's perspective.

> * Change doesn't work:
>   When you click on Change, the box remains on Confirm Password:
>   ![image](https://private-user-images.githubusercontent.com/214579/364829046-bcbe1107-e5a5-480a-8297-50b7c0ce10cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODI5MDQ2LWJjYmUxMTA3LWU1YTUtNDgwYS04Mjk3LTUwYjdjMGNlMTBjYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYjg1ODk3MGRjZjFkMTdhYTU0MzYzYWMwODNjNmE4ZjkxNmEwOGVmMjZlMGE5N2NmMGE1NWE1YzFhNjVkMTljJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.99X14lf7zFGVG0_tUiDFyl5_NtrNqICzyxLdn4DediE)

That's due to the cosmetic bug above. (It should be on password.)

>   You click on Verify, it then goes to Password.
>   The Change Password procedure on client is:
>   After click on Change, you have to type your current PIN
>   ![image](https://private-user-images.githubusercontent.com/214579/364830014-9298dcf0-8b06-4f77-a22b-e03771ff971d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODMwMDE0LTkyOThkY2YwLThiMDYtNGY3Ny1hMjJiLWUwMzc3MWZmOTcxZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMTE2MDM2MTM4MWJlYjExMzQ2Y2Q5NzBiNWZkN2VlM2E3ZDI5Yzg3ZTMzNDNjNTQ2MjIzMGY5MDU3NzYyYzc5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.KBp-vbqneZ70VEckQjXY5lZtwfudb9YniejhigriGio)
>   And then you click on Verify, New Password is now accessible:
>   ![image](https://private-user-images.githubusercontent.com/214579/364830224-aa2a09f2-5029-4d3b-a1ac-d04348423c3c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODMwMjI0LWFhMmEwOWYyLTUwMjktNGQzYi1hMWFjLWQwNDM0ODQyM2MzYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YTdhYWU3ZWNlNjEzNGE3MjUyODE5N2IxM2QwZGJkZjlmY2M5NmEzMmZlMTllMDlhM2M4MDJmNjg1Y2E5NGE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.CWnDqfAjxuPGonj06xPny-IcIMdn9_jDfIxQZfcEGDY)
>   Click on Verify again and then Confirm Password is accessible and Confirm button appears:
>   ![image](https://private-user-images.githubusercontent.com/214579/364831074-916f1e9e-3f00-464d-8182-0cd9f43074b6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODMxMDc0LTkxNmYxZTllLTNmMDAtNDY0ZC04MTgyLTBjZDlmNDMwNzRiNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04YWQ4ZTQ4MjJkYjc1N2NkYjZlNjFhNzEzYzcwMjM2ZjY3YjNiNWVjMWNmNTRlNjJjMmE3NDU5MDQ2MjRiNmNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Kx1sPfOLomQS_aqdWV0HCjE-4TLLuhEPjcyLwRUceGQ)
>   When all information are filled in, if you click on Reset, The procedure will revert back to Password:
>   ![image](https://private-user-images.githubusercontent.com/214579/364830865-7d88a90f-60d0-4f7d-9ef1-992b94cbaf71.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODMwODY1LTdkODhhOTBmLTYwZDAtNGY3ZC05ZWYxLTk5MmI5NGNiYWY3MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZmJlZTBmMmNjNTQ2MDFlOGJlNTY0NWQ0NTUyNWYwZWM0NDNhNjEyMjExNzM2NjU4OTA2MWEyY2Y5OTY0ZjU3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.8eJVMTcbENHKXQRHeOqRg-3LhzhJq6AMdOfo7B74Puw)

That's due to there being no field selection code in the JS handler. I.e. If you click on the password box it doesn't do anything.

That's because I didn't write any code to handle changing the current input field from the GUI. It's done programmatically based on which buttons the user clicks on and which response packets are sent from the login server.

Reasons being:
  1) The responses from the server can reset the window.
  2) Filtering non-numeric input is a pain. (And subject to a permanent ban on official servers. It will even crash some clients, and persist between runs....)
  3) I wanted to get the login itself working before focusing too much on cosmetics.

> * NEW PIN Code:
>   It's buggy but works for new accounts.
>   On client it's just like change minus to enter the current PIN code yet the box is white, not blue:

Cosmetic bug strikes again. That's the field select changing what input field is active. (It doesn't have any code to handle the input field that isn't used for a the new password flow.)

>   Enter New Password and then click on Verify.
>   ![image](https://private-user-images.githubusercontent.com/214579/364834246-cb992be4-a039-4745-82c5-68829c54969b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODM0MjQ2LWNiOTkyYmU0LWEwMzktNDc0NS04MmM1LTY4ODI5YzU0OTY5Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMDQ0YzNlYzM3NDk3ZDkzZTQ4ZGRhYjRlMjZlNDA1ZTNjZWY3MmY3YzE0OTcwYTU5YmNjNzAwNjgzMzMzYmZhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-r4eBu76YOzu-EJMFeE001XjVKyjoPFJjZhWzfjQegw)
>   Enter Confirm Password and then click on Confirm.
>   ![image](https://private-user-images.githubusercontent.com/214579/364834406-bb731321-3f75-4028-ad2d-353dc8e3fe5b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODM0NDA2LWJiNzMxMzIxLTNmNzUtNDAyOC1hZDJkLTM1M2RjOGUzZmU1Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MDgzYzhmZjhlMzI5OTg5M2NiOWY0NjkyNDA5ZTg4OWIxNjRiNzhjZDc5M2RmMjY5NjE2NzgwMjg5N2FiYzNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bGT-tA80DjJIyqcIK7si4dTkV02gBm1l8KsM3FB6_zc)
> 
> Browser also get 3 notifications box overlapped: ![image](https://private-user-images.githubusercontent.com/214579/364835209-162cc280-de0e-40c5-8575-7fd236955378.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNjkwMzIsIm5iZiI6MTcyNjA2ODczMiwicGF0aCI6Ii8yMTQ1NzkvMzY0ODM1MjA5LTE2MmNjMjgwLWRlMGUtNDBjNS04NTc1LTdmZDIzNjk1NTM3OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxNTMyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Njc3OTcxZTdlODdjYWJjMDk2NjNhMGQxODliNTFkODczYjc1ODAzNmVjNjI2NTRhYWQ5YWY0ZDQ1YzM5MDA1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.1uBJpeZdWPkWwVt13cJGXz1po_XZvX_ohK412ITgzDE)

Those message windows should be "modal" windows. I.e. (You shouldn't be able to do anything but click on their OK button when they are on the screen.) The PincodeWindow doesn't have a disable flag to block input when a message is displayed. So you can get multiple message windows if you don't close them. (Something else that needs to be fixed.)

All that being said, the original issue of "not working Packetver: 20130618" is fixed. Maybe we should close this bug and open new ones for the other stuff.

**Comment by redpolline** [comment]
> Same issue here. Can't make new pincode and can't login either. This original ui is pretty unintuitive with the lower case "change" to begin with,

Hey, that's graphics pulled from the grf. I didn't translate them. :stuck_out_tongue_winking_eye: 

Fun fact if you highlight the change button without clicking on it, it becomes a "_cha**r**ge_" button instead. :smile: 

> but the random fileds activating and the 3 popups make it a real a chaos :D I don't even know what I am supposed to do :D

Yeah, I need to clean it up. Although to be fair, the official client doesn't tell you much either, and I got lost the first time I used it. 

The normal input flow is still there however. So if your used to inputting a Pincode in the official client, and able to ignore the graphical glitches in the current UI, you'll still complete the process successfully. Even if that process doesn't look right. But some people can't handle the graphics being wrong. (I would know, it's the reason why I set confirm password as the default field to begin with. Having the top field being the input field for the default login process didn't look right to me. Of course, that was before I knew the correct process from the official client.... :blush:)

**Comment by MrAntares** [comment]
I know the UI looking stupid is not your fault. The translation project needs some hardcore supervision because the translations are sometimes really bad and there are alot of typos and wrong letter casing. But the original UIs are bad as well. Non-responsive 90's tech. They just look bad and feel bad and not intuitive. It's a shame we have to mimic them. At least we can have plugins that improve on these. Thx for the fix anyways!

@AtomixRO Can you test the fix sometime?

**Comment by MrAntares** [comment]
There is still an error. in `HC.SECOND_PASSWD_LOGIN` When status is 0, it either means the pincode is correct OR the feature is disables and you must proceed with the login. When this happens and there is no other pincode packet received before this, then the pincode window is not even loaded and it starts referencing unloaded objects resulting in loads of errors.

**Comment by MrAntares** [comment]
I don't really get why `PincodeWindow.selectInput` is always called with 0 in the `CharEngine.js` and why the values are manupilated later in the UI itself, but I guess the issue is somewhere in this logic. It enables/disables the wrong fields

**Comment by MrAntares** [comment]
There is another issue. in `CharEngine.js` the `onConnectionAccepted` is used for:
- `ACCEPT_ENTER_NEO_UNION_HEADER`
- `ACCEPT_ENTER_NEO_UNION`
- `ACCEPT_ENTER_NEO_UNION_LIST`

Packet `ACCEPT_ENTER_NEO_UNION` is automatically followed by `SECOND_PASSWD_LOGIN` if the feature is enabled. This is count number 1 for the packet. But since there is a direct request for pincode in the `onConnectionAccepted` function using the `sendPincodeRequest` function it also requests for an additional pincode for bot the `ACCEPT_ENTER_NEO_UNION_HEADER` and the `ACCEPT_ENTER_NEO_UNION` packets and possibly the `ACCEPT_ENTER_NEO_UNION_LIST` wherever it is used. This results in receiving the `SECOND_PASSWD_LOGIN` packet 3 or 4 times back to back and this is what creates the triplicated message popups. I will remove this call for additional pincodes, but please revise how this should work, because there is some misunderstanding here or the logic is not the same for every packet version.

**Comment by redpolline** [comment]
> I don't really get why `PincodeWindow.selectInput` is always called with 0 in the `CharEngine.js` and why the values are manupilated later in the UI itself, but I guess the issue is somewhere in this logic. It enables/disables the wrong fields

It shouldn't be called externally. If it is that's a bug. (It would be a leftover from an earlier revision of the code. In reality, selectInput() shouldn't even be exported from PincodeWindow at all, and nothing external should be calling it.)

As for the zero arg bit, that was just my way of saying "reset to initial input field." (Which helps given I need to change the field order. I'll send a commit later for that.)

>there is some misunderstanding here or the logic is not the same for every packet version.

Yes, based on the tests I made, the older packetver expects the client to request the SECOND_PASSWD_LOGIN packet from the server. In those packetvers if we don't send that request to the server, then the pincode window doesn't get initialized properly.

That functionality should be gated by a packetver check however. So, I'm not sure why it would be triggering on packetvers that don't need it. It doesn't on my end with the tests I've made, but then again I don't have proper full clients to test those packetvers on my end. (The patch I sent was based on a frankinclient made with a 2021 grf against an rAthena built with the older packetver. I can't test with an original client.)

**Comment by MrAntares** [comment]
Thx! Gonna test tmr

**Comment by AtomixRO** [comment]
I tested the update. @redpolline 

- Login: Works as intended.

- New Pin Code: Works more easily but isn't like on client as you go straight to Password field and type the password once instead as mentioned before.

- Change password doesn't work, same issue as before.

**Comment by AtomixRO** [comment]
Another issue which is an important cosmetic matter is that whenever players create a new pin code that is banned from the server because it's too easy such as same number like 1111 0000 or even 1234 (maybe some others which are stored server side)

You then get the following error message (which is confusing)
![image](https://github.com/user-attachments/assets/f0673495-6c7b-4586-b099-283a618d0df0)

instead of the following on client:
![image](https://github.com/user-attachments/assets/42a37f4d-eab7-4df8-916c-28b4d0d40e9d)

**Comment by MrAntares** [comment]
@AtomixRO can you plz test this again? There were quite a few updates
